File: sdks/typescript/src/mqttTransport.ts
/*
 * Copyright 2026 ACP Project
 * Licensed under the Apache License, Version 2.0
 * See LICENSE file for details.
 */

import mqtt, { IClientOptions, MqttClient } from "mqtt";
import { readFileSync } from "node:fs";
import { JsonMap, JsonValue } from "./jsonSupport.js";
import { invalidArgument, transportError, validationError } from "./errors.js";
import {
  AuthConfig,
  AuthType,
  assertAllowedAuthTypes,
  authParameter,
  parseAuthConfig,
  parseAuthFromService,
  serializeAuthConfig
} from "./transportAuth.js";

export const DEFAULT_MQTT_QOS = 1;
export const DEFAULT_MQTT_TOPIC_PREFIX = "acp/agent";

export type MqttMessageHandler = (message: JsonMap) => boolean | Promise<boolean>;
const MQTT_AUTH_TYPES: ReadonlySet<AuthType> = new Set(["none", "username_password", "mtls", "custom"]);

function toQos(qos: number): 0 | 1 | 2 {
  if (qos <= 0) {
    return 0;
  }
  if (qos >= 2) {
    return 2;
  }
  return 1;
}

function clampQos(qos: number): number {
  return Math.min(2, Math.max(0, qos));
}

function pickString(service: JsonMap | undefined, key: string, fallback: string): string {
  const value = service?.[key];
  if (typeof value === "string" && value.trim()) {
    return value.trim();
  }
  return fallback;
}

function valueAsNumber(value: JsonValue | undefined): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "string" && value.trim()) {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) {
      return parsed;
    }
  }
  return undefined;
}

export function metadataProperties(message: JsonMap): Record<string, string> {
  const envelope =
    message.envelope && typeof message.envelope === "object" && !Array.isArray(message.envelope)
      ? (message.envelope as JsonMap)
      : {};
  const metadata: Record<string, string> = {};
  for (const [source, destination] of [
    ["acp_version", "acp_version"],
    ["message_class", "acp_message_class"],
    ["message_id", "acp_message_id"],
    ["operation_id", "acp_operation_id"],
    ["sender", "acp_sender"]
  ]) {
    const value = envelope[source];
    if (typeof value === "string" && value.trim()) {
      metadata[destination] = value.trim();
    }
  }
  return metadata;
}

export class MqttTransportClient {
  public readonly broker_url: string;
  public readonly qos: number;
  public readonly topic_prefix: string;
  public readonly timeout_seconds: number;
  public readonly keepalive_seconds: number;
  public readonly auth: AuthConfig | undefined;

  public constructor(
    brokerUrl: string,
    qos = DEFAULT_MQTT_QOS,
    topicPrefix = DEFAULT_MQTT_TOPIC_PREFIX,
    timeoutSeconds = 10,
    keepaliveSeconds = 30,
    auth?: unknown
  ) {
    if (!brokerUrl.trim()) {
      throw invalidArgument("broker_url must be provided");
    }
    this.broker_url = brokerUrl.trim();
    this.qos = clampQos(qos);
    this.topic_prefix = topicPrefix.trim().replace(/\/+$/, "") || DEFAULT_MQTT_TOPIC_PREFIX;
    this.timeout_seconds = Math.max(1, timeoutSeconds);
    this.keepalive_seconds = Math.max(5, keepaliveSeconds);
    this.auth = parseAuthConfig(auth);
    assertAllowedAuthTypes(this.auth, MQTT_AUTH_TYPES, "MQTT transport");
  }

  public static agentIdentifierToken(agentId: string): string {
    const match = /^agent:(?<name>[^@]+)(?:@(?<domain>.+))?$/.exec(agentId);
    if (!match?.groups?.name) {
      throw validationError(`Invalid agent identifier: ${agentId}`);
    }
    const base = match.groups.domain ? `${match.groups.name}.${match.groups.domain}` : match.groups.name;
    const normalized = base
      .split("")
      .map((char) => (/^[A-Za-z0-9._-]$/.test(char) ? char : "."))
      .join("")
      .split(".")
      .filter((segment) => segment.length > 0)
      .join(".")
      .toLowerCase();
    return normalized || "unknown";
  }

  public static topicForAgent(agentId: string, topicPrefix = DEFAULT_MQTT_TOPIC_PREFIX): string {
    return `${topicPrefix.replace(/\/+$/, "")}/${MqttTransportClient.agentIdentifierToken(agentId)}`;
  }

  public static buildServiceHint(
    agentId: string,
    brokerUrl: string,
    topic?: string,
    qos = DEFAULT_MQTT_QOS,
    topicPrefix = DEFAULT_MQTT_TOPIC_PREFIX,
    auth?: unknown
  ): JsonMap {
    const hint: JsonMap = {
      broker_url: brokerUrl,
      topic: topic?.trim() || MqttTransportClient.topicForAgent(agentId, topicPrefix),
      qos: clampQos(qos)
    };
    const parsedAuth = parseAuthConfig(auth);
    assertAllowedAuthTypes(parsedAuth, MQTT_AUTH_TYPES, "MQTT transport");
    const serialized = serializeAuthConfig(parsedAuth);
    if (serialized) {
      hint.auth = serialized;
    }
    return hint;
  }

  private connectClient(
    brokerUrl: string,
    auth: AuthConfig | undefined
  ): Promise<MqttClient> {
    const options: IClientOptions = {
      protocolVersion: 5,
      keepalive: this.keepalive_seconds,
      reconnectPeriod: 0,
      connectTimeout: this.timeout_seconds * 1000
    };
    const parsed = new URL(brokerUrl);
    if (parsed.username) {
      options.username = decodeURIComponent(parsed.username);
      options.password = decodeURIComponent(parsed.password);
    }
    if (auth && auth.type !== "none") {
      if (auth.type === "username_password") {
        options.username = authParameter(auth, "username", "MQTT username_password auth");
        options.password = authParameter(auth, "password", "MQTT username_password auth");
      } else if (auth.type === "custom") {
        if (auth.parameters.username?.trim()) {
          options.username = auth.parameters.username.trim();
          options.password = (auth.parameters.password ?? "").trim();
        }
      }
      if (auth.type === "mtls") {
        options.cert = readFileSync(authParameter(auth, "cert_path", "MQTT mTLS auth"));
        options.key = readFileSync(authParameter(auth, "key_path", "MQTT mTLS auth"));
        if (auth.parameters.ca_path?.trim()) {
          options.ca = readFileSync(auth.parameters.ca_path.trim());
        }
      } else if (
        auth.type === "custom" &&
        auth.parameters.cert_path?.trim() &&
        auth.parameters.key_path?.trim()
      ) {
        options.cert = readFileSync(auth.parameters.cert_path.trim());
        options.key = readFileSync(auth.parameters.key_path.trim());
        if (auth.parameters.ca_path?.trim()) {
          options.ca = readFileSync(auth.parameters.ca_path.trim());
        }
      }
    }
    return new Promise((resolve, reject) => {
      const client = mqtt.connect(brokerUrl, options);
      const onConnect = (): void => {
        cleanup();
        resolve(client);
      };
      const onError = (error: Error): void => {
        cleanup();
        client.end(true);
        reject(error);
      };
      const cleanup = (): void => {
        client.off("connect", onConnect);
        client.off("error", onError);
      };
      client.once("connect", onConnect);
      client.once("error", onError);
    });
  }

  public async publish(message: JsonMap, recipientAgentId: string, service?: JsonMap): Promise<void> {
    const brokerUrl = pickString(service, "broker_url", this.broker_url);
    const auth = parseAuthFromService(service) ?? this.auth;
    assertAllowedAuthTypes(auth, MQTT_AUTH_TYPES, "MQTT transport");
    const topic = pickString(
      service,
      "topic",
      MqttTransportClient.topicForAgent(recipientAgentId, this.topic_prefix)
    );
    const qos = clampQos(valueAsNumber(service?.qos) ?? this.qos);
    const properties = metadataProperties(message);
    const client = await this.connectClient(brokerUrl, auth);
    try {
      await new Promise<void>((resolve, reject) => {
        client.publish(
          topic,
          JSON.stringify(message),
          {
            qos: toQos(qos),
            properties: {
              userProperties: properties
            }
          },
          (error?: Error) => {
            if (error) {
              reject(error);
              return;
            }
            resolve();
          }
        );
      });
    } catch (error) {
      throw transportError(`mqtt publish failed: ${String(error)}`);
    } finally {
      client.end(true);
    }
  }

  public async consume(
    agentId: string,
    handler: MqttMessageHandler,
    service?: JsonMap,
    maxMessages = 0,
    pollTimeoutMs = 1000
  ): Promise<number> {
    const brokerUrl = pickString(service, "broker_url", this.broker_url);
    const auth = parseAuthFromService(service) ?? this.auth;
    assertAllowedAuthTypes(auth, MQTT_AUTH_TYPES, "MQTT transport");
    const topic = pickString(service, "topic", MqttTransportClient.topicForAgent(agentId, this.topic_prefix));
    const qos = clampQos(valueAsNumber(service?.qos) ?? this.qos);
    const limit = maxMessages === 0 ? Number.MAX_SAFE_INTEGER : maxMessages;
    const client = await this.connectClient(brokerUrl, auth);
    let processed = 0;
    try {
      await new Promise<void>((resolve, reject) => {
        client.subscribe(topic, { qos: toQos(qos) }, (error: Error | null) => {
          if (error instanceof Error) {
            reject(error);
            return;
          }
          resolve();
        });
      });
      await new Promise<void>((resolve) => {
        const timeout = setTimeout(() => {
          cleanup();
          resolve();
        }, pollTimeoutMs);
        const cleanup = (): void => {
          clearTimeout(timeout);
          client.off("message", onMessage);
        };
        const onMessage = async (_topic: string, payload: Buffer): Promise<void> => {
          if (processed >= limit) {
            cleanup();
            resolve();
            return;
          }
          try {
            const parsed = JSON.parse(payload.toString("utf-8")) as JsonValue;
            if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
              await handler(parsed as JsonMap);
            }
          } catch {
            // Keep behavior tolerant: invalid messages are ignored.
          }
          processed += 1;
          if (processed >= limit) {
            cleanup();
            resolve();
          }
        };
        client.on("message", onMessage);
      });
      return processed;
    } catch (error) {
      throw transportError(`mqtt consume failed: ${String(error)}`);
    } finally {
      client.end(true);
    }
  }
}

File: sdks/go/mqtt_transport.go
/*
 * Copyright 2026 ACP Project
 * Licensed under the Apache License, Version 2.0
 * See LICENSE file for details.
 */

package acp

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

const (
	DefaultMQTTQoS         = 1
	DefaultMQTTTopicPrefix = "acp/agent"
)

type MQTTMessageHandler func(message map[string]any) bool

type MqttTransportClient struct {
	BrokerURL        string
	QoS              int
	TopicPrefix      string
	TimeoutSeconds   int
	KeepaliveSeconds int
	Auth             *AuthConfig
}

var mqttAgentPattern = regexp.MustCompile(`^agent:(?P<name>[^@]+)(?:@(?P<domain>.+))?$`)

func NewMqttTransportClient(brokerURL string, qos int, topicPrefix string, timeoutSeconds int, keepaliveSeconds int) (*MqttTransportClient, error) {
	return NewMqttTransportClientWithAuth(brokerURL, qos, topicPrefix, timeoutSeconds, keepaliveSeconds, nil)
}

func NewMqttTransportClientWithAuth(
	brokerURL string,
	qos int,
	topicPrefix string,
	timeoutSeconds int,
	keepaliveSeconds int,
	auth *AuthConfig,
) (*MqttTransportClient, error) {
	if strings.TrimSpace(brokerURL) == "" {
		return nil, InvalidArgument("broker_url must be provided")
	}
	if strings.TrimSpace(topicPrefix) == "" {
		topicPrefix = DefaultMQTTTopicPrefix
	}
	normalizedAuth, err := NormalizeAuthConfig(auth)
	if err != nil {
		return nil, err
	}
	return &MqttTransportClient{
		BrokerURL:        strings.TrimSpace(brokerURL),
		QoS:              clampMQTTQoS(qos),
		TopicPrefix:      strings.TrimRight(strings.TrimSpace(topicPrefix), "/"),
		TimeoutSeconds:   maxInt(timeoutSeconds, 1),
		KeepaliveSeconds: maxInt(keepaliveSeconds, 5),
		Auth:             normalizedAuth,
	}, nil
}

func clampMQTTQoS(qos int) int {
	if qos < 0 {
		return 0
	}
	if qos > 2 {
		return 2
	}
	return qos
}

func MQTTAgentIdentifierToken(agentID string) (string, error) {
	matches := mqttAgentPattern.FindStringSubmatch(agentID)
	if len(matches) == 0 {
		return "", ValidationError(fmt.Sprintf("Invalid agent identifier: %s", agentID))
	}
	var name, domain string
	for index, group := range mqttAgentPattern.SubexpNames() {
		switch group {
		case "name":
			name = matches[index]
		case "domain":
			domain = matches[index]
		}
	}
	base := name
	if domain != "" {
		base = name + "." + domain
	}
	var normalized strings.Builder
	for _, char := range strings.ToLower(base) {
		if (char >= 'a' && char <= 'z') ||
			(char >= '0' && char <= '9') ||
			char == '.' || char == '_' || char == '-' {
			normalized.WriteRune(char)
		} else {
			normalized.WriteRune('.')
		}
	}
	parts := []string{}
	for _, part := range strings.Split(normalized.String(), ".") {
		part = strings.TrimSpace(part)
		if part != "" {
			parts = append(parts, part)
		}
	}
	if len(parts) == 0 {
		return "unknown", nil
	}
	return strings.Join(parts, "."), nil
}

func MQTTTopicForAgent(agentID string, topicPrefix string) (string, error) {
	token, err := MQTTAgentIdentifierToken(agentID)
	if err != nil {
		return "", err
	}
	prefix := strings.TrimSpace(topicPrefix)
	if prefix == "" {
		prefix = DefaultMQTTTopicPrefix
	}
	return strings.TrimRight(prefix, "/") + "/" + token, nil
}

func BuildMQTTServiceHint(agentID, brokerURL, topic string, qos int, topicPrefix string) (map[string]any, error) {
	return BuildMQTTServiceHintWithAuth(agentID, brokerURL, topic, qos, topicPrefix, nil)
}

func BuildMQTTServiceHintWithAuth(
	agentID, brokerURL, topic string, qos int, topicPrefix string, auth *AuthConfig,
) (map[string]any, error) {
	if strings.TrimSpace(topic) == "" {
		derived, err := MQTTTopicForAgent(agentID, topicPrefix)
		if err != nil {
			return nil, err
		}
		topic = derived
	}
	normalizedAuth, err := NormalizeAuthConfig(auth)
	if err != nil {
		return nil, err
	}
	hint := map[string]any{
		"broker_url": strings.TrimSpace(brokerURL),
		"topic":      strings.TrimSpace(topic),
		"qos":        clampMQTTQoS(qos),
	}
	if authMap := AuthConfigToMap(normalizedAuth); authMap != nil {
		hint["auth"] = authMap
	}
	return hint, nil
}

func MQTTMetadataProperties(message map[string]any) map[string]string {
	out := map[string]string{}
	envelope, _ := message["envelope"].(map[string]any)
	for source, destination := range map[string]string{
		"acp_version":   "acp_version",
		"message_class": "acp_message_class",
		"message_id":    "acp_message_id",
		"operation_id":  "acp_operation_id",
		"sender":        "acp_sender",
	} {
		raw, _ := envelope[source].(string)
		trimmed := strings.TrimSpace(raw)
		if trimmed != "" {
			out[destination] = trimmed
		}
	}
	return out
}

func mqttStringValue(service map[string]any, key string, fallback string) string {
	if service == nil {
		return fallback
	}
	value, ok := service[key].(string)
	if ok && strings.TrimSpace(value) != "" {
		return strings.TrimSpace(value)
	}
	return fallback
}

func mqttNumberValue(service map[string]any, key string, fallback int) int {
	if service == nil {
		return fallback
	}
	switch typed := service[key].(type) {
	case int:
		return clampMQTTQoS(typed)
	case float64:
		return clampMQTTQoS(int(typed))
	default:
		return fallback
	}
}

func (client *MqttTransportClient) connect(brokerURL string, auth *AuthConfig) (mqtt.Client, error) {
	options := mqtt.NewClientOptions()
	options.AddBroker(brokerURL)
	options.SetProtocolVersion(5)
	options.SetKeepAlive(time.Duration(client.KeepaliveSeconds) * time.Second)
	options.SetConnectTimeout(time.Duration(client.TimeoutSeconds) * time.Second)
	options.SetAutoReconnect(false)
	if parsed, err := url.Parse(brokerURL); err == nil {
		if parsed.User != nil {
			password, _ := parsed.User.Password()
			options.SetUsername(parsed.User.Username())
			options.SetPassword(password)
		}
	}
	if auth != nil {
		switch auth.Type {
		case "none":
		case "username_password":
			username, err := RequireAuthParameter(auth, "username", "MQTT username_password auth")
			if err != nil {
				return nil, err
			}
			password, err := RequireAuthParameter(auth, "password", "MQTT username_password auth")
			if err != nil {
				return nil, err
			}
			options.SetUsername(username)
			options.SetPassword(password)
		case "custom":
			if username := strings.TrimSpace(auth.Parameters["username"]); username != "" {
				options.SetUsername(username)
				options.SetPassword(strings.TrimSpace(auth.Parameters["password"]))
			}
		case "mtls":
			tlsConfig, err := buildMQTTTLSConfig(auth, true)
			if err != nil {
				return nil, err
			}
			options.SetTLSConfig(tlsConfig)
		default:
			return nil, ValidationError(fmt.Sprintf("MQTT transport does not support auth type: %s", auth.Type))
		}
		if auth.Type == "custom" {
			if certPath := strings.TrimSpace(auth.Parameters["cert_path"]); certPath != "" {
				tlsConfig, err := buildMQTTTLSConfig(auth, false)
				if err != nil {
					return nil, err
				}
				options.SetTLSConfig(tlsConfig)
			}
		}
	}
	mqttClient := mqtt.NewClient(options)
	token := mqttClient.Connect()
	if !token.WaitTimeout(time.Duration(client.TimeoutSeconds) * time.Second) {
		return nil, TransportError("mqtt connect timeout")
	}
	if token.Error() != nil {
		return nil, TransportError(fmt.Sprintf("mqtt connect failed: %v", token.Error()))
	}
	return mqttClient, nil
}

func (client *MqttTransportClient) Publish(message map[string]any, recipientAgentID string, service map[string]any) error {
	brokerURL := mqttStringValue(service, "broker_url", client.BrokerURL)
	auth, err := effectiveServiceAuth(service, client.Auth)
	if err != nil {
		return err
	}
	topic := mqttStringValue(service, "topic", "")
	if topic == "" {
		derived, err := MQTTTopicForAgent(recipientAgentID, client.TopicPrefix)
		if err != nil {
			return err
		}
		topic = derived
	}
	qos := mqttNumberValue(service, "qos", client.QoS)
	body, err := json.Marshal(message)
	if err != nil {
		return TransportError(fmt.Sprintf("unable to serialize MQTT payload: %v", err))
	}
	mqttClient, err := client.connect(brokerURL, auth)
	if err != nil {
		return err
	}
	defer mqttClient.Disconnect(50)
	token := mqttClient.Publish(topic, byte(qos), false, body)
	if !token.WaitTimeout(time.Duration(client.TimeoutSeconds) * time.Second) {
		return TransportError("mqtt publish timeout")
	}
	if token.Error() != nil {
		return TransportError(fmt.Sprintf("mqtt publish failed: %v", token.Error()))
	}
	return nil
}

func (client *MqttTransportClient) Consume(
	agentID string,
	handler MQTTMessageHandler,
	service map[string]any,
	maxMessages int,
	pollTimeout time.Duration,
) (int, error) {
	brokerURL := mqttStringValue(service, "broker_url", client.BrokerURL)
	auth, err := effectiveServiceAuth(service, client.Auth)
	if err != nil {
		return 0, err
	}
	topic := mqttStringValue(service, "topic", "")
	if topic == "" {
		derived, err := MQTTTopicForAgent(agentID, client.TopicPrefix)
		if err != nil {
			return 0, err
		}
		topic = derived
	}
	qos := mqttNumberValue(service, "qos", client.QoS)
	if maxMessages <= 0 {
		maxMessages = int(^uint(0) >> 1)
	}
	if pollTimeout <= 0 {
		pollTimeout = time.Second
	}
	mqttClient, err := client.connect(brokerURL, auth)
	if err != nil {
		return 0, err
	}
	defer mqttClient.Disconnect(50)
	processed := 0
	done := make(chan struct{}, 1)
	token := mqttClient.Subscribe(topic, byte(qos), func(_ mqtt.Client, message mqtt.Message) {
		if processed >= maxMessages {
			select {
			case done <- struct{}{}:
			default:
			}
			return
		}
		var parsed map[string]any
		if err := json.Unmarshal(message.Payload(), &parsed); err == nil && parsed != nil {
			_ = handler(parsed)
		}
		processed++
		if processed >= maxMessages {
			select {
			case done <- struct{}{}:
			default:
			}
		}
	})
	if !token.WaitTimeout(time.Duration(client.TimeoutSeconds) * time.Second) {
		return processed, TransportError("mqtt subscribe timeout")
	}
	if token.Error() != nil {
		return processed, TransportError(fmt.Sprintf("mqtt subscribe failed: %v", token.Error()))
	}
	select {
	case <-done:
	case <-time.After(pollTimeout):
	}
	return processed, nil
}

func buildMQTTTLSConfig(auth *AuthConfig, requireClientCertificate bool) (*tls.Config, error) {
	tlsConfig := &tls.Config{MinVersion: tls.VersionTLS12}
	if caPath := strings.TrimSpace(auth.Parameters["ca_path"]); caPath != "" {
		data, err := os.ReadFile(caPath)
		if err != nil {
			return nil, ValidationError(fmt.Sprintf("unable to read auth.parameters.ca_path: %v", err))
		}
		pool := x509.NewCertPool()
		if ok := pool.AppendCertsFromPEM(data); !ok {
			return nil, ValidationError("unable to parse CA bundle from auth.parameters.ca_path")
		}
		tlsConfig.RootCAs = pool
	}
	certPath := strings.TrimSpace(auth.Parameters["cert_path"])
	keyPath := strings.TrimSpace(auth.Parameters["key_path"])
	if requireClientCertificate {
		if certPath == "" || keyPath == "" {
			return nil, ValidationError("MQTT mTLS auth requires auth.parameters.cert_path and auth.parameters.key_path")
		}
	}
	if certPath != "" && keyPath != "" {
		certificate, err := tls.LoadX509KeyPair(certPath, keyPath)
		if err != nil {
			return nil, ValidationError(fmt.Sprintf("unable to load MQTT client certificate: %v", err))
		}
		tlsConfig.Certificates = []tls.Certificate{certificate}
	}
	return tlsConfig, nil
}
